k4GaudiPandora/include/DDPandoraPFANewAlgorithm.h

```diff
@@ -0,0 +1,342 @@
+   *  @param  v0Collections the Vertex collections of V0s
+   *  @param  trackerHitLinkCollections the associations between trackerHits and simTrackerHits
+   *  @param  trackCollections collections of tracks
+   *  @param  eCalCollections  CalorimeterHit Collection for the ECal
+   *  @param  hCalCollections  CalorimeterHit Collection for the HCal
+   *  @param  mCalCollections  CalorimeterHit Collection for the Muon Calo
+   *  @param  lCalCollections  CalorimeterHit Collection for the LCal
+   *  @param  lhCalCollections CalorimeterHit Collection for the HLCal
+   *  @param  caloLinkCollections the associations between CalorimeterHits and MCParticles
+   * 
+   *  @return tuple of reconstructed: (Clusters, RecoParticles, Verticies)
+   *  
+   */
+  std::tuple<edm4hep::ClusterCollection, edm4hep::ReconstructedParticleCollection, edm4hep::VertexCollection> operator()(
+   const std::vector<const edm4hep::MCParticleCollection*>& MCParticleCollections,
+   const std::vector<const edm4hep::VertexCollection*>& kinkCollections,
+   const std::vector<const edm4hep::VertexCollection*>& prongCollections,
+   const std::vector<const edm4hep::VertexCollection*>& splitCollections,
+   const std::vector<const edm4hep::VertexCollection*>& v0Collections,
+   const std::vector<const edm4hep::TrackCollection*>& trackCollections,
+   const std::vector<const edm4hep::TrackerHitSimTrackerHitLinkCollection*>& trackerHitLinkCollections,
+   const std::vector<const edm4hep::CalorimeterHitCollection*>& eCalCollections,
+   const std::vector<const edm4hep::CalorimeterHitCollection*>& hCalCollections,
+   const std::vector<const edm4hep::CalorimeterHitCollection*>& mCalCollections,
+   const std::vector<const edm4hep::CalorimeterHitCollection*>& lCalCollections,
+   const std::vector<const edm4hep::CalorimeterHitCollection*>& lhCalCollections,
+   const std::vector<const edm4hep::CaloHitSimCaloHitLinkCollection*>& caloLinkCollections
+ ) const override;
+
+
+
+  /**
+     *  @brief  End, called at shutdown
+     */
+  StatusCode finalize();
+
+  /**
+     *  @brief  Get address of the pandora instance
+     *
+     *  @return address of the pandora instance
+     */
+  const pandora::Pandora* GetPandora() const;
+
+  /**
+     *  @brief  Get address of the current lcio event
+     *
+     *  @param  pPandora address of the relevant pandora instance
+     *
+     *  @return address of the current lcio event
+     */
+  //static const EVENT::LCEvent* GetCurrentEvent(const pandora::Pandora* const pPandora);
+
+private:
+  /**
+     *  @brief  Register user algorithm factories, energy correction functions and particle id functions,
+     *          insert user code here
+     */
+  pandora::StatusCode RegisterUserComponents() const;
+
+  /**
+     *  @brief  Copy some steering parameters between settings objects
+     */
+  void FinaliseSteeringParameters();
+
+  /**
+     *  @brief  Reset the pandora pfa new processor
+     */
+  void Reset() const;
```

Comment:
The presence of a `Reset` function immediately makes me assume that this is not re-entrant and can, thus not run in parallel. I am not sure if a functional algorithm can tweak the necessary property (@jmcarcell ?) to communicate that to the scheduler.

For a non re-entrant algorithm, I would almost lean towards implementing it as a regular `Gaudi::Algorithm` simply to not give the wrong impression. The preferred long-term solution would obviously be to make this re-entrant, but I am not sure yet if that is easily possible.